File: hacking_with_perl/debug_endpoints.pl
#!/usr/bin/perl

use strict;
use warnings;
use LWP::UserAgent;
use Term::ANSIColor;

# Function to test if a debug endpoint exists
sub test_debug_endpoint {
    my $url = shift;

    my @debug_endpoints = (
        "debug",
        "debug.php",
        "debug.jsp",
        "debug.aspx",
        "debug.txt",
        "debug.log",
        "trace",
        "trace.php",
        "trace.jsp",
        "trace.aspx",
        "trace.txt",
        "trace.log",
        "info",
        "info.php",
        "info.jsp",
        "info.aspx",
        "info.txt",
        "info.log"
    );

    my $ua = LWP::UserAgent->new;
    $ua->timeout(10);

    foreach my $endpoint (@debug_endpoints) {
        my $target_url = $url . "/" . $endpoint;
        my $response = $ua->get($target_url);

        if ($response->is_success) {
            print colored("[+] Vulnerable: $target_url\n", 'green');
            return 1;
        }
    }

    print colored("[-] Not vulnerable: No debug endpoints found\n", 'red');
    return 0;
}

# Main program
sub main {
    print "Enter your target website url: ";
    my $target_website = <STDIN>;
    chomp $target_website;

    print "Testing for Exposed Debug Endpoints on $target_website...\n";

    if (test_debug_endpoint($target_website)) {
        print "\nTo further test the vulnerability, try accessing the debug endpoints listed above in a web browser or using a tool like curl.\n";
    }
}

main();

File: hacking_with_go/debug_endpoints.go
package main

import (
	"fmt"
	"net/http"
	"strings"
)

// Color codes for output formatting
const (
	Reset  = "\033[0m"
	Red    = "\033[31m"
	Green  = "\033[32m"
	Yellow = "\033[33m"
)

// Function to check if a URL is vulnerable to exposed debug endpoints
func checkDebugEndpoints(url string) bool {
	// List of common debug endpoints
	debugEndpoints := []string{
		"/debug",
		"/debug/vars",
		"/debug/pprof",
		"/debug/pprof/heap",
		"/debug/pprof/goroutine",
		"/debug/pprof/block",
		"/debug/pprof/threadcreate",
		"/trace",
		"/debug/fgprof",
		"/debug/log",
		"/debug/zip",
		"/debug/mutex",
		"/debug/allocs",
	}

	// Iterate over each debug endpoint
	for _, endpoint := range debugEndpoints {
		// Construct the full URL
		fullURL := url + endpoint

		// Send a GET request to the endpoint
		resp, err := http.Get(fullURL)
		if err != nil {
			fmt.Printf("Error accessing %s: %v\n", fullURL, err)
			continue
		}

		// Check if the response status code indicates success (200 OK)
		if resp.StatusCode == http.StatusOK {
			fmt.Printf("%s is %svulnerable%s to exposed debug endpoint: %s\n", url, Red, Reset, endpoint)
			return true
		}

		// Close the response body
		resp.Body.Close()
	}

	// If no debug endpoints were found to be exposed
	fmt.Printf("%s is %snot vulnerable%s to exposed debug endpoints\n", url, Green, Reset)
	return false
}

func main() {
	fmt.Println("Welcome to the Exposed Debug Endpoints Tester")

	// Prompt user to enter target website URL
	var targetURL string
	fmt.Print("Enter your target website URL: ")
	fmt.Scanln(&targetURL)

	// Check if the target URL starts with "http://" or "https://", if not, add it
	if !strings.HasPrefix(targetURL, "http://") && !strings.HasPrefix(targetURL, "https://") {
		targetURL = "http://" + targetURL
	}

	// Check for exposed debug endpoints
	vulnerable := checkDebugEndpoints(targetURL)

	if vulnerable {
		fmt.Println("\nTo test this vulnerability, try accessing the debug endpoints listed above.")
		fmt.Println("For example, you can try visiting", targetURL+"/debug/vars")
	}

	fmt.Println("\nProgram execution completed.")
}
